Create hyperlinks on PDF export

alvarogonzalezsotillo opened the pull request:
Related with https://github.com/xournalpp/xournalpp/issues/1827, I added a simple URL recognition in text boxes.

A text box with only one line starting with **xxxx:** is interpreted as a URL, and the exported PDF has a hyperlink to it.

In order to maintain the previous behaviour, I have added a new setting. I didn't found any suitable tab, so I created a new one

![2020-10-15-193053_950x557_scrot](https://user-images.githubusercontent.com/1097236/96166386-35253b00-0f1e-11eb-9e1f-85730da493a8.png)

I still need some advice:

- I can't figure out a simple/elegant way to get the `Settings` pointer from the `TextView` or `DocumentView` class, so I create a new `Settings` in the `DocumentView` constructor. Do you think this is acceptable?

- I can translate the new setting to spanish (and maybe other languages with the help of an automatic translator), but I don't know how to handle `.po` and `.pot` files.

Technius commented:
Sorry for leaving this alone, I've been really busy. I'll take a look over the weekend.

alvarogonzalezsotillo commented:
This is an open source project and it is supposed to be funny, which implies no deadlines.
Have a nice weekend!

Technius reviewed (changes requested):
Thanks for the PR! We get to make our own deadlines ;) More seriously, it would be a big waste of everybody's effort and time to let PRs sit until they're outdated. And, well, I want this feature too.

> I can't figure out a simple/elegant way to get the Settings pointer from the TextView or DocumentView class, so I create a new Settings in the DocumentView constructor. Do you think this is acceptable?

This is probably out-of-sync with the global settings, so we'll need a different way to handle this. A better place to pass in the new setting is through `src/pdf/base/XojCairoPdfExport.cpp`.

> I can translate the new setting to spanish (and maybe other languages with the help of an automatic translator), but I don't know how to handle .po and .pot files. I suppose `xgettext` is invoked from cmake somehow, but I need some instructions.

The preferred way to add translations is through [Crowdin](https://crowdin.com/project/xournalpp), which you can do after the PR is merged.

alvarogonzalezsotillo commented on src/view/TextView.cpp:
```diff
@@ -39,19 +41,39 @@ void TextView::updatePangoFont(PangoLayout* layout, const Text* t) {
+      // Assume url is escaped, change only apostrophe
+      uri += regex_replace(str,std::regex("'"), "%27"); 
+      uri += "'";
+      cairo_tag_begin(cr, CAIRO_TAG_LINK, uri.c_str());
```
I will surround the URL export code in conditional compilation, like:

```c++
#if` CAIRO_VERSION >= CAIRO_VERSION_ENCODE(1, 16, 0)
  ......
#endif
```
There are two options
- Show the export hyperlinks options, but disabled
- Hide completely the export hiperlink tab (my preferred)
What solution do you prefer?

alvarogonzalezsotillo commented:
I will surround the URL export code in conditional compilation, like:

```c++
#if` CAIRO_VERSION >= CAIRO_VERSION_ENCODE(1, 16, 0)
  ......
#endif
```
There are two options
- Show the export hyperlinks options, but disabled
- Hide completely the export hiperlink tab (my preferred)

What solution do you prefer?

Technius commented:
I think a good place for the export hyperlinks option would be in the `Load / Save` tab.

alvarogonzalezsotillo commented:
Sorry for the late response.

I have modified some classes in order to get the `Config` in the `XojCairoPdfExport`, so I can get the `Settings` in `DocumentView::drawElement`.


![diagrama](https://user-images.githubusercontent.com/1097236/100480930-fa085100-30f2-11eb-8aed-6e79d6e3a056.png)


I have changed quite a few files, and I don' feel confident it is the best solution. Also, I am getting some runtime exceptions from time to time, and I have no idea why.

Do you think this is the right aproach?

Technius reviewed (commented):
Sorry for the late reply, I haven't had any time this past month.

I think we've both been looking at this the wrong way. URI information should be part of the model, not part of the rendering. Thus, the best solution is to check when the text contents are a URI in `Text::setText`. We can easily do this using [`g_uri_parse`](https://developer.gnome.org/glib/stable/glib-URI-Functions.html#g-uri-parse), and we just need to store the resulting `GUri*` in `Text`. Then in `TextView::drawText`, we just need to check if the Text is indeed a URI; if so, then we can convert the URI with `g_uri_to_string` and then add the cairo tags.

This will also make it easy to add custom links in the future.
